pkg/miniogw/gateway-storj.go

```diff
@@ -247,6 +255,39 @@ func (s *storjObjects) MakeBucketWithLocation(ctx context.Context,
 	return err
 }
 
+func (s *storjObjects) CopyObject(ctx context.Context, srcBucket, srcObject, destBucket,
+	destObject string, srcInfo minio.ObjectInfo) (objInfo minio.ObjectInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	rr, err := s.getObject(ctx, srcBucket, srcObject)
+	if err != nil {
+		return objInfo, err
+	}
+
+	defer utils.LogClose(rr)
+
+	r, err := rr.Range(ctx, 0, srcInfo.Size)
+	if err != nil {
+		return objInfo, err
+	}
+
+	defer utils.LogClose(r)
+
+	return s.putObject(ctx, destBucket, destObject, r, srcInfo)
+}
+
+func (s *storjObjects) putObject(ctx context.Context, bucket, object string, r io.ReadCloser,
+	srcInfo minio.ObjectInfo) (objInfo minio.ObjectInfo, err error) {
+	defer mon.Task()(&ctx)(&err)
+
+	hr, err := hash.NewReader(r, srcInfo.Size, "", "")
+	if err != nil {
+		return objInfo, err
+	}
+
+	return s.PutObject(ctx, bucket, object, hr, srcInfo.UserDefined)
```

Comment:
The idea of proposing this `putObject` helper method was to avoid calling the `s.PutObject()`, but instead call the Bucket/Object Store API directly.

There are 2 issues with calling `s.PutObject`:
  1. We need to create that extra `hash.NewReader` wrap, which does additional work - calculates some hashes. We have no use of those hashes, so this is just a waste of CPU usage and we want to avoid it.
  2. We are losing the Content Type info. It needs to be combined with `srcInfo.UserDefined` before calling `s.PutObject`, but this is again not optimal.

Instead this helper method should look like:
```
func (s *storjObjects) putObject(ctx context.Context, bucket, object string, io.Reader, meta objects.SerializableMeta) (objInfo minio.ObjectInfo, err error) {
  // copy here the relevant part of code from s.PutObject
}
```